pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java

```diff
@@ -335,6 +346,27 @@ public void healthCheck(@Suspended AsyncResponse asyncResponse,
                 });
     }
 
+    private void checkDeadlockedThreads() {
+        ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+        long[] threadIds = threadBean.findDeadlockedThreads();
+        if (threadIds != null && threadIds.length > 0) {
+            ThreadInfo[] threadInfos = threadBean.getThreadInfo(threadIds, false, false);
+            String threadNames = Arrays.stream(threadInfos)
+                    .map(threadInfo -> threadInfo.getThreadName() + "(tid=" + threadInfo.getThreadId() + ")").collect(
+                            Collectors.joining(", "));
+            if (System.currentTimeMillis() - threadDumpLoggedTimestamp
+                    > LOG_THREADDUMP_INTERVAL_WHEN_DEADLOCK_DETECTED) {
+                threadDumpLoggedTimestamp = System.currentTimeMillis();
+                LOG.error("Deadlocked threads detected. {}\n{}", threadNames,
+                        ThreadDumpUtil.buildThreadDiagnosticString());
+            } else {
+                LOG.error("Deadlocked threads detected. {}", threadNames);
+            }
+            throw new IllegalStateException("Deadlocked threads detected. " + threadNames);
```

Comment:
A small problem, Whenever a deadlock is found, is it a bit violent to directly judge that the broker is unhealthy?

Reply:
Can you explain why/how you categorize deadlock as a small problem?

Deadlock is an unrecoverable state for threads on the JVM, and the presence of deadlocked threads means something is wrong/stuck within the broker. I agree with @lhotari that we should consider deadlocked brokers to be unhealthy.

Reply:
Sorry, Not expressed clearly. I do not think this is a small problem.

Just wanted to confirm the following conclusions:

> Deadlock is an unrecoverable state for threads on the JVM, and the presence of deadlocked threads means something is wrong/stuck within the broker. I agree with @lhotari that we should consider deadlocked brokers to be unhealthy.